README.rst
======
pytill
======


.. image:: https://img.shields.io/pypi/v/pytill.svg
        :target: https://pypi.python.org/pypi/pytill

.. image:: https://img.shields.io/travis/manifoldco/pytill.svg
        :target: https://travis-ci.org/manifoldco/pytill

.. image:: https://readthedocs.org/projects/pytill/badge/?version=latest
        :target: https://pytill.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


Python package for using Till Mobile brought to you by https://manifold.co


* Free software: BSD license
* Documentation: https://pytill.readthedocs.io.


How to use
--------

Set up a plan for till and retrieve till username and api key. This packages expects the username and api key are injected as env vars ``USERNAME`` and ``API_KEY``.

You can do this simply through manifold: (see: https://docs.manifold.co/docs/cli-quickstart-6JMEw1CD6wguwIYymUuAQ6)

.. code-block:: bash

    manifold create -p till-project --product till --plan free # provision a free till resource
    manifold run -p till-project -- python # inject vars


.. code-block:: python

    import pytill # or from pytill import pytill

    # send a message
    pytill.send_message(['19024880000'], 'I am sending a till message isnt that cool!')

    # ask a question
    # note that asking a question is how we open two-way communicate in Till
    # so this is also how you intiate listening to responses to a sms number with Till
    question = pytill.make_question('How cool is Till mobile?', 'my-question', 'my.webhook/listens/here')
    pytill.send_question(['19024441111', '16139094888'], [question],  'my-project')

    # retrieve result
    print(pytill.get_results(question_tag='my-question'))

    # example output

    # {'meta': {'limit': 20, 'next': None, 'offset': 0, 'previous': None, 'total_count': 1}, 'objects': [{'created': '2018-08-27T19:07:35.423855', 'guid': '252cd98f-5969-44c9-a955-7bb54e6f0d19', 'origin_phone_number': '+16508668969', 'participant_guid': '883c8f57-74b9-43cb-bb72-c7634b97651a', 'participant_phone_number': '+19024000158', 'project_launch_guid': '80262aea-a77e-4a0c-911f-23b959aea6da', 'project_launch_participant_guid': '8a23ee27-4841-4f2b-83d5-2a59a05825b8', 'project_tag': 'my-project', 'question_display_order': '0', 'question_guid': '8c8c168f-87d5-454d-8bae-09781312c097', 'question_tag': 'my-question', 'question_text': 'How cool is Till mobile?', 'result_answer': 'really cool!', 'result_guid': '252cd98f-5969-44c9-a955-7bb54e6f0d19', 'result_response': 'really cool!', 'result_timestamp': '2018-08-27T19:07:35.423855', 'updated': '2018-08-27T19:07:35.423884'}]}

    # retrive stats about till usage
    print(pytill.get_stats())


HISTORY.rst
=======
History
=======

1.0.0 (2018-08-31)
------------------

* Release full package.

0.1.0 (2018-08-27)
------------------

* First release on PyPI.


tests/test_pytill.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `pytill` package."""
from unittest.mock import Mock

import pytest
import requests

from pytill import pytill

GUID = 'cool_guid'
STATS_URL = 'sweet-url/dog'
RESULTS_URL = 'nice-results/bro'

@pytest.fixture
def send_success(monkeypatch):
    def mock_post(url, data=None, json=None, **kwargs):
        msg_response = {
            'project_launch_guid': GUID,
            'stats_url': STATS_URL,
            'results_url': RESULTS_URL
        }
        mock = Mock()
        mock.request_for_status = Mock()
        mock.json = Mock(return_value=msg_response)
        return mock
    monkeypatch.setattr(requests, 'post', mock_post)

@pytest.fixture
def send_failure(monkeypatch):
    def mock_post(url, data=None, json=None, **kwargs):
        def mock_raise():
            raise requests.HTTPError('oops')
        mock = Mock()
        mock.request_for_status = Mock(side_effect=mock_raise())
        return mock

    monkeypatch.setattr(requests, 'post', mock_post)

@pytest.fixture
def get_failure(monkeypatch):
    def mock_get(url, params=None, **kwargs):
        def mock_raise():
            raise requests.HTTPError('oops')
        mock = Mock()
        mock.request_for_status = Mock(side_effect=mock_raise())
        return mock

    monkeypatch.setattr(requests, 'get', mock_get)

@pytest.fixture
def get_result(monkeypatch):
    def mock_get(url, params=None, **kwargs):
        msg_response = {'something': 'somethingelse'}
        mock = Mock()
        mock.request_for_status = Mock()
        mock.json = Mock(return_value=msg_response)
        return mock
    monkeypatch.setattr(requests, 'get', mock_get)

def test_send_message(send_success):
    expected_guid = GUID
    r = pytill.send_message(['19021234567'],
        'Hi how are you? Its dom I am doing a test',
        tag='Greeting', voice=True
    )
    assert expected_guid == r['project_launch_guid']
       
def test_send_message_raises(send_failure):
    with pytest.raises(requests.HTTPError):
        pytill.send_message(['19021234567'], 'Hi how are you? Its dom I am doing a test')

def test_make_question():
    expected_question =  {
          'conclude_on': 'Thanks for answering.',
          'responses': ['yes', 'no'],
          'tag': 'cool-q-1',
          'text': 'This is my cool question?',
          'webhook': 'sweet-web-hook-bro'
    }
    question = pytill.make_question(
        "This is my cool question?", "cool-q-1", "sweet-web-hook-bro",
         responses=["yes", "no"], conclude_on="Thanks for answering."
    )
    assert expected_question == question

def test_send_question(send_success):
    expected_response = {
        'project_launch_guid': GUID,
        'stats_url': STATS_URL,
        'results_url': RESULTS_URL
    }
    questions = [pytill.make_question(
        'How are you?', "how-r-u", "how-r-u-webhook.com",
        responses=["Good", "Bad"], conclude_on="Thanks"
    )]
    r = pytill.send_question(['19021234567'],
        questions,
        "my-project-tag",
        introduction="Hey can I ask you a question",
        conclusion="Thanks for letting me ask you a question"
    )
    assert expected_response == r
      
def test_send_question_raises(send_failure):
    with pytest.raises(requests.HTTPError):
        questions = [pytill.make_question('How are you?', "how-r-u", "how-r-u-webhook.com")]
        pytill.send_question(['190212345678'], 'greetings-project', questions)

def test_get_results(get_result):
    expected_response = {'something': 'somethingelse'}
    r = pytill.get_results()
    assert expected_response == r

def test_get_results_raises(get_failure):
    with pytest.raises(requests.HTTPError):
        pytill.get_results()

def test_get_result(get_result):
    expected_response = {'something': 'somethingelse'}
    r = pytill.get_result('some-guid')
    assert expected_response == r

def test_get_result_raises(get_failure):
    with pytest.raises(requests.HTTPError):
        pytill.get_results()

def test_get_stats(get_result):
    expected_response = {'something': 'somethingelse'}
    r = pytill.get_stats()
    assert expected_response == r

def test_get_stats_raises(get_failure):
    with pytest.raises(requests.HTTPError):
        pytill.get_stats()

def test_get_stat(get_result):
    expected_response = {'something': 'somethingelse'}
    r = pytill.get_stat('some-guid')
    assert expected_response == r

def test_get_stat_raises(get_failure):
    with pytest.raises(requests.HTTPError):
        pytill.get_stat('some-guid')


pytill/pytill.py
# pylint: disable=too-many-arguments
# -*- coding: utf-8 -*-

"""Main module."""
import os

import requests

TILL_USERNAME = os.getenv("USERNAME")
TILL_API_KEY = os.getenv("API_KEY")
TILL_SEND_URL = "https://platform.tillmobile.com/api/send?username={}&api_key={}"
TILL_RESULTS_URL = "https://platform.tillmobile.com/api/results"
TILL_RESULT_URL = "https://platform.tillmobile.com/api/results/{}/"
TILL_STATS_URL = "https://platform.tillmobile.com/api/stats/"
TILL_STAT_URL = "https://platform.tillmobile.com/api/stats/{}/"


def send_message(numbers, text, tag=None, voice=False):
    """Send a message with till.

    Args:
        numbers (list): List of numbers to send message to.
        text (str): Message to send.
        tag (:obj:`str`, optional): Tag to tag the project with. Defaults to None.
        voice (:obj:`bool`, optional): Whether or not to send message as boice. Defaults to None.

    Returns:
        dict: Data relating to monitoring the message sent:
        .. codeblock:: python
            {
                'stats_url': param1,
                'results_url': param2,
                'project_launch_guid': param3
            }

    Raises:
        HTTPError: For various reasons due to misconfiguration or insuffecient plan.

    Examples:
        Send a voice message with tag greeting to multiple people.
 
        >>> resp = pytill.send_message(
            ['19021231234', '19024564567'],
            text='Hi!', tag='Greeting', voice=True)
        {'stats_url': param1, 'results_url': param2, 'project_launch_guid': param3}
    """
    msg = {
        'phone': numbers,
        'text': text
    }

    if voice:
        msg['method'] = "VOICE"

    if tag:
        msg['tag'] = tag

    resp = requests.post(
        TILL_SEND_URL.format(
            TILL_USERNAME,
            TILL_API_KEY
        ),
        json=msg
    )

    resp.raise_for_status()
    return resp.json()


def make_question(text, tag, webhook, responses=None, conclude_on=None):
    """Make a question to ask.

    Args:
        text (str): Question to ask.
        tag (str): Question tag for retrieving results.
        webhook (str): Webhook to listen for results on.
        responses (:obj:`list`, optional): List of responses to provide. Defaults to None.
        conclude_on (:obj:`str`, optional): Response the question should conclude on. Defaults to None.

    Returns:
        dict: Question to be asked
    
    Examples:
        >>> question = pytill.make_question(
            "This is my cool question?", "cool-q-1", "sweet-web-hook-bro",
            responses=["yes", "no"], conclude_on="Thanks for answering."
        )
    """
    question = {
        "text": text,
        "tag": tag,
        "webhook": webhook
    }
    if conclude_on:
        question["conclude_on"] = conclude_on

    if responses:
        question["responses"] = responses
    return question


def send_question(numbers, questions, tag, introduction=None, conclusion=None, voice=False):
    """Send questions to ask

    Note that asking a question is how we open two-way communicate in Till
    so this is also how you intiate listening to responses to a sms number with Till

    Args:
        numbers (list): List of numbers to send questions to.
        questions (list): List of questions to send.
        tag (str): Project tag for retrieving results for various questions.
        introduction (:obj:`str`, optional): Optional introduction to question(s). Defaults to None.
        conclusion (:obj:`str`, optional): Optional conclusion to question(s). Defaults to None.
        voice (:obj:`bool`, optional): Whether or not to ask the questions via voice. Defaults to None.

    Returns:
        dict: Data relating to monitoring the question sent:
            {
                'stats_url': param1,
                'results_url': param2,
                'project_launch_guid': param3
            }
    
    Example:
        >>> question = pytill.make_question('How cool is Till mobile?', 'my-question', 'my.webhook/listens/here')
        >>> pytill.send_question(['19024441111', '16139094888'], [question],  'my-project')

    """
    question_payload = {
        "tag": tag,
        "phone": numbers,
        "questions": questions
    }
    if introduction:
        question_payload["introduction"] = introduction

    if conclusion:
        question_payload["conclusion"] = conclusion

    if voice:
        question_payload["method"] = "VOICE"

    resp = requests.post(
        TILL_SEND_URL.format(
            TILL_USERNAME,
            TILL_API_KEY
        ),
        json=question_payload
    )

    resp.raise_for_status()
    return resp.json()


def get_results(question_tag=None, project_tag=None, project_launch_guid=None,
                participant_guid=None):
    """ Get results for questions asked
    
    Args:
        question_tag (:obj:`str`, optional): Tag from question to retreive on. Defaults to None.
        project_tag (:obj:`str`, optional): Tag from project (ie. list of questions to retreive on. Defaults to None.
        project_launch_guid (:obj:`str`, optional): Project guid to retreive on. Defaults to None.
        participant_guid (:obj:`str`, optional): Participant guid (ie. person/number) to retreive on. Defaults to None.

    Example:
        >>> pytill.get_results(question_tag='my-question')
        {'meta': {'limit': 20, 'next': None, 'offset': 0, 'previous': None, 'total_count': 1}, 'objects': [{'created': '2018-08-27T19:07:35.423855', 'guid': '252cd98f-5969-44c9-a955-7bb54e6f0d19', 'origin_phone_number': '+16508668969', 'participant_guid': '883c8f57-74b9-43cb-bb72-c7634b97651a', 'participant_phone_number': '+19024000158', 'project_launch_guid': '80262aea-a77e-4a0c-911f-23b959aea6da', 'project_launch_participant_guid': '8a23ee27-4841-4f2b-83d5-2a59a05825b8', 'project_tag': 'my-project', 'question_display_order': '0', 'question_guid': '8c8c168f-87d5-454d-8bae-09781312c097', 'question_tag': 'my-question', 'question_text': 'How cool is Till mobile?', 'result_answer': 'really cool!', 'result_guid': '252cd98f-5969-44c9-a955-7bb54e6f0d19', 'result_response': 'really cool!', 'result_timestamp': '2018-08-27T19:07:35.423855', 'updated': '2018-08-27T19:07:35.423884'}]}
    
    """
    query = {
        'username': TILL_USERNAME,
        'api_key': TILL_API_KEY
    }
    if question_tag:
        query['question_tag'] = question_tag
    if project_tag:
        query['project_tag'] = project_tag
    if project_launch_guid:
        query['project_launch_guid'] = project_launch_guid
    if participant_guid:
        query['participant_guid'] = participant_guid

    resp = requests.get(TILL_RESULTS_URL, params=query)
    resp.raise_for_status()
    return resp.json()


def get_result(result_guid):
    """Get single result from result guid retreived in a ``get_results`` call"""
    query = {
        'username': TILL_USERNAME,
        'api_key': TILL_API_KEY
    }
    resp = requests.get(TILL_RESULT_URL.format(result_guid), params=query)
    resp.raise_for_status()
    return resp.json()


def get_stats(project_tag=None):
    """Get status of your Till usage. This can be broken down by project_tag."""
    query = {
        'username': TILL_USERNAME,
        'api_key': TILL_API_KEY
    }
    if project_tag:
        query['project_tag'] = project_tag

    resp = requests.get(TILL_STATS_URL, params=query)
    resp.raise_for_status()
    return resp.json()


def get_stat(project_launch_guid):
    """Get particulat status for a project_launch_guid"""
    query = {
        'username': TILL_USERNAME,
        'api_key': TILL_API_KEY
    }
    resp = requests.get(TILL_STAT_URL.format(project_launch_guid), params=query)
    resp.raise_for_status()
    return resp.json()
